### DISCUSSION TITLE
How to access the users "account_token" and display it?

### DISCUSSION THREAD

**Discussion by MatD1** [nextauthjs/next-auth#2752] [Help] [answered]
I've setup a quick app to try and access the users access_token that is grabbed once the user is logged in and then display it on the home page. This app will only be used for myself but I need the access token later on to use within the API.

Image on the whats happening [Imgur](https://i.imgur.com/EvDEvNv.jpg)

Seriously any help would be amazing!! 
If you'd like to see the app this is the github link: https://github.com/MatD1/matsites3

**Comment by balazsorban44** [comment, marked as answer]
https://next-auth.js.org/configuration/callbacks

**Reply by MatD1** [reply]
Sorry for the late reply on this but I am still not able to show the users token? I have tried using {session.user.token}, {session.user.accessToken}, {session.token}, none of these are working, These where the ways I was able to previously display the tokens. Is it different in V4? or do I need to pass another prop with the useSession?

**Reply by balazsorban44** [reply]
make sure you check the jwt callback as well. it's the same in v4

**Reply by MatD1** [reply]
Yeah I have managed to console log the token and it works perfectly. The only issue is I think im passing it to my index.js wrong, it's just returning "account is undefined"
```
import { useSession, signIn, signOut } from "next-auth/react"

export default function Home() {
  const { data: session, account } = useSession()
  if ( session) {
    return (
      <>
        <h1>Signed in as {session.user.email}</h1> <br />
        <h1>{session.user.name}</h1>
        <h1>{account.access_token}</h1>
        <button onClick={() => signOut()}>Sign out</button>
      </>
    )
  }
  return (
    <>
      Not signed in <br />
      <button onClick={() => signIn()}>Sign in</button>
    </>
  )
}
```

**Reply by balazsorban44** [reply]
there is no account on session. you have to forward it. have another read at the docs

**Reply by MatD1** [reply]
For the life of me I cannot figure this out, I have read the docs numerous times and please keep in mind something that seems so easy to you is not so much to me I'm still learning a lot about this as I go. Of course your help has been great though. 

Thats my callback and the console log in session is returning nothing. I swear I've tried a number of things but I cannot figure it out, I do apologise for the annoyance. 
```js
 callbacks: {
    async jwt({ token, account, session }) {
      if (account) {
        token.accessToken = account.access_token
        console.log("JWT Token", account.access_token)
      }
      return token
    },
    async session({ session, token, user, account }) {
      if (account) {
      session.accessToken = token.accessTokenn
      access_token = session.accessToken
      console.log("Session Token", session.accessToken)
      }
      return session
    }
```

**Reply by balazsorban44** [reply]
I understand. We hope that docs help most people, and if they aren't, we are open to suggestions, what is unclear. So if you are struggling to understand it, please tell me why. I am not sure how to make this more clear in the docs currently.

The correct code in your case:
```js
 callbacks: {
    async jwt({ token, account }) { // There is nothing called `session` here
      if (account) {
        token.accessToken = account.access_token
      }
      return token
    },
    async session({ session, token }) { // There is nothing called `account` or `user` in your case as documented. `user` is only present if you had used a database persisted session.
      session.accessToken = token.accessTokenn
      return session
    }
```

will be available as `useSession().data?.accessToken`. What the `session` callback returns is effectively the value of `data` in the client.

Actually, this is the exact code snippet the docs shows you:
https://next-auth.js.org/configuration/callbacks#jwt-callback
![](https://i.imgur.com/W83w5IE.png)

https://next-auth.js.org/configuration/callbacks#session-callback
![](https://i.imgur.com/2An69A3.png)

So I am genuinely curious which part of the docs wasn't clear? I would like to make it better if possible.

**Reply by MatD1** [reply]
Okay I don't understand what I've really changed but it's working now, I've only changed it back to what the docs said and of course it works but I had that yesterday and it wasn't working, maybe I had a spelling mistake or something. But a huge thanks to you for the help I much appreciate it very much. 

In my opinion the docs are great just a little hard for someone who isn't a pro but I understand that considering what this is working with! Overall this project is great and has worked so much better than other solutions I have tried. 

I now need to try and use this in an API.

**Reply by balazsorban44** [reply]
Nice to hear! 

TypeScript and/or ESLint would probably both be nice tools to avoid spelling mistakes, at least TS has a learning curve, so I understand if it's not on your roadmap yet. All in all, I am happy that things worked out! 🎉
